src/ray/core_worker/reference_count.cc

```diff
@@ -164,4 +315,339 @@ ReferenceCounter::GetAllReferenceCounts() const {
+  }
+  if (!it->second.owner.has_value() &&
+      borrower_ref.contained_in_borrowed_id.has_value()) {
+    // We don't have owner information about this object ID yet and the worker
+    // received it because it was nested in another ID that the worker was
+    // borrowing. Copy this information to our local table.
+    RAY_CHECK(borrower_ref.owner.has_value());
+    AddBorrowedObjectInternal(object_id, *borrower_it->second.contained_in_borrowed_id,
+                              borrower_ref.owner->first, borrower_ref.owner->second);
+  }
+  std::vector<rpc::WorkerAddress> new_borrowers;
+
+  // The worker is still using the reference, so it is still a borrower.
+  if (borrower_ref.RefCount() > 0) {
+    auto inserted = it->second.borrowers.insert(worker_addr).second;
+    // If we are the owner of id, then send WaitForRefRemoved to borrower.
+    if (inserted) {
+      RAY_LOG(DEBUG) << "Adding borrower " << worker_addr.ip_address << " to id "
+                     << object_id;
+      new_borrowers.push_back(worker_addr);
+    }
+  }
+
+  // Add any other workers that this worker passed the ID to as new borrowers.
+  for (const auto &nested_borrower : borrower_ref.borrowers) {
+    auto inserted = it->second.borrowers.insert(nested_borrower).second;
+    if (inserted) {
+      RAY_LOG(DEBUG) << "Adding borrower " << nested_borrower.ip_address << " to id "
+                     << object_id;
+      new_borrowers.push_back(nested_borrower);
+    }
+  }
+
+  // If we own this ID, then wait for all new borrowers to reach a ref count
+  // of 0 before GCing the object value.
+  if (it->second.owned_by_us) {
+    for (const auto &addr : new_borrowers) {
+      WaitForRefRemoved(it, addr);
+    }
+  }
+
+  // Recursively merge any references that were contained in this object, to
+  // handle any borrowers of nested objects.
+  for (const auto &inner_id : borrower_ref.contains) {
+    MergeRemoteBorrowers(inner_id, worker_addr, borrowed_refs);
+  }
+}
+
+void ReferenceCounter::WaitForRefRemoved(const ReferenceTable::iterator &ref_it,
+                                         const rpc::WorkerAddress &addr,
+                                         const ObjectID &contained_in_id) {
+  const ObjectID &object_id = ref_it->first;
+  rpc::WaitForRefRemovedRequest request;
+  // Only the owner should send requests to borrowers.
+  RAY_CHECK(ref_it->second.owned_by_us);
+  request.mutable_reference()->set_object_id(object_id.Binary());
+  request.mutable_reference()->set_owner_id(ref_it->second.owner->first.Binary());
+  request.mutable_reference()->mutable_owner_address()->CopyFrom(
+      ref_it->second.owner->second);
+  request.set_contained_in_id(contained_in_id.Binary());
+  request.set_intended_worker_id(addr.worker_id.Binary());
+
+  auto it = borrower_cache_.find(addr);
+  if (it == borrower_cache_.end()) {
+    RAY_CHECK(client_factory_ != nullptr);
+    it = borrower_cache_.emplace(addr, client_factory_(addr.ToProto())).first;
```

Comment:
Would be great to share this RPC factory/cache logic with other parts of the core worker

Reply:
I think that will have to be part of a much larger refactor.